unmarshaler.go

```diff
@@ -129,213 +165,369 @@ func (d *decoder) arrayIndex(shouldAppend bool, v reflect.Value) int {
 
-// scopeWithKey performs target scoping when unmarshaling an ast.KeyValue node.
-//
-// The goal is to hop from target to target recursively using the names in key.
-// Parts of the key should be used to resolve field names for structs, and as
-// keys when targeting maps.
-//
-// When encountering slices, it should always use its last element, and error
-// if the slice does not have any.
-func (d *decoder) scopeWithKey(x target, key ast.Iterator) (target, bool, error) {
-	var (
-		err   error
-		found bool
-	)
-
-	for key.Next() {
-		n := key.Node()
+func (d *decoder) handleKeyPart(key ast.Iterator, v reflect.Value, nextFn handlerFn, makeFn valueMakerFn) (reflect.Value, error) {
+	var rv reflect.Value
 
-		x, found, err = d.scopeTableTarget(false, x, string(n.Data))
-		if err != nil || !found {
-			return nil, found, err
+	// First, dispatch over v to make sure it is a valid object.
+	// There is no guarantee over what it could be.
+	switch v.Kind() {
+	case reflect.Map:
+		// Create the key for the map element. For now assume it's a string.
+		mk := reflect.ValueOf(string(key.Node().Data))
+
+		// If the map does not exist, create it.
+		if v.IsNil() {
+			v = reflect.MakeMap(v.Type())
+			rv = v
 		}
-	}
 
-	return x, true, nil
-}
+		mv := v.MapIndex(mk)
+		set := false
+		if !mv.IsValid() {
+			// If there is no value in the map, create a new one according to
+			// the map type. If the element type is interface, create either a
+			// map[string]interface{} or a []interface{} depending on whether
+			// this is the last part of the array table key.
+
+			t := v.Type().Elem()
+			if t.Kind() == reflect.Interface {
+				mv = makeFn()
+			} else {
+				mv = reflect.New(t).Elem()
+			}
+			set = true
+		} else if mv.Kind() == reflect.Interface {
+			mv = mv.Elem()
+			if !mv.IsValid() {
+				mv = makeFn()
+			}
+			set = true
+		}
 
-//nolint:cyclop
-// scopeWithArrayTable performs target scoping when unmarshaling an
-// ast.ArrayTable node.
-//
-// It is the same as scopeWithKey, but when scoping the last part of the key
-// it creates a new element in the array instead of using the last one.
-func (d *decoder) scopeWithArrayTable(x target, key ast.Iterator) (target, bool, error) {
-	var (
-		err   error
-		found bool
-	)
+		x, err := nextFn(key, mv)
+		if err != nil {
+			return reflect.Value{}, err
+		}
 
-	for key.Next() {
-		n := key.Node()
-		if !n.Next().Valid() { // want to stop at one before last
-			break
+		if x.IsValid() {
+			mv = x
+			set = true
 		}
 
-		x, found, err = d.scopeTableTarget(false, x, string(n.Data))
-		if err != nil || !found {
-			return nil, found, err
+		if set {
+			v.SetMapIndex(mk, mv)
+		}
+	case reflect.Struct:
+		f, found := structField(v, string(key.Node().Data))
+		if !found {
+			d.skipUntilTable = true
+			return reflect.Value{}, nil
 		}
-	}
 
-	n := key.Node()
+		x, err := nextFn(key, f)
+		if err != nil || d.skipUntilTable {
+			return reflect.Value{}, err
+		}
+		if x.IsValid() {
+			f.Set(x)
```

Comment:
Wonder if this is still called more often than before the changes.

Reply:
That's 5*24 = 120 bytes copied.

Reply:
After more digging, it seems like I'm wrong: `Set` is called only when elements are appended to slices when unmarshalling array tables.